### PR TITLE
Lanzatool: generate custom os-release

### PR DESCRIPTION
**Pull request by nikstur**
Generate a custom os-release to provide nice boot menu entries for systemd-boot.

### CODE REVIEW

File: rust/lanzatool/src/generation.rs [hunk 1]
```diff
@@ -94,6 +90,33 @@ impl fmt::Display for Generation {
...
+        .file_name()
+        .and_then(|x| x.to_str())
+        .context("Failed to convert path to filename string.")?;
+
+    Ok(String::from(file_name))
+}
+
+fn read_build_time(toplevel: &Path) -> Result<String> {
+    Ok(
+        time::OffsetDateTime::from_unix_timestamp(fs::metadata(toplevel)?.mtime())?
```
Comment (RaitoBezarius) [inline rust/lanzatool/src/generation.rs:114]: Is this even reliable?

Comment (nikstur) [reply rust/lanzatool/src/generation.rs:114]: Why not? It's the same stuff the systemdboot builder does. We could do a default value (e.g ."Unknown").

File: rust/lanzatool/src/generation.rs [hunk 2]
```diff
@@ -51,41 +47,45 @@ impl Generation {
...
     }
 
     pub fn is_specialized(&self) -> Option<SpecialisationName> {
         self.specialisation_name.clone()
     }
+
+    /// Describe the generation in a single line.
+    ///
+    /// This is currently implemented by poking around the filesystem to find the necessary data.
+    /// Ideally, the needed data should be included in the bootspec.
```
Comment (blitz) [inline rust/lanzatool/src/generation.rs:74]: Should this be feedback we give to the bootspec RFC?

File: rust/lanzatool/src/generation.rs [hunk 3]
```diff
@@ -51,41 +47,45 @@ impl Generation {
...
+        let toplevel = &self.spec.bootspec.toplevel.0;
+
+        let nixos_version = fs::read_to_string(toplevel.join("nixos-version"))
+            .unwrap_or_else(|_| String::from("Unknown"));
+        let kernel_version =
+            read_kernel_version(toplevel).context("Failed to read kernel version.")?;
+        let build_time = read_build_time(toplevel).unwrap_or_else(|_| String::from("Unknown"));
+
+        Ok(format!(
+            "Generation {} NixOS {}, Linux Kernel {}, Built on {}",
```
Comment (blitz) [inline rust/lanzatool/src/generation.rs:85]: I guess you want to generation first for ordering? If so, that would be a good comment here.

**Review by blitz — APPROVED** [review]
Looks good to me. I left some nits.

File: rust/lanzatool/src/generation.rs (hunk 2 above)
Comment (nikstur) [reply rust/lanzatool/src/generation.rs:74]: I spoke to @RaitoBezarius about this. I think he wants to address that at the next bootspec meeting.

File: rust/lanzatool/src/generation.rs (hunk 1 above)
Comment (nikstur) [reply rust/lanzatool/src/generation.rs]: Made this default to "Unknown". If we realize this is an error we can change it later.

**Pull request merged**